File: packages/contracts/src/agents.ts
/**
 * Mirrors services/api/app/modules/agents/schemas.py.
 */
import { z } from "zod";

export const AgentCreate = z.object({
  name: z.string().min(1).max(150),
  business_identity: z.string().min(1).max(200),
  description: z.string().nullable().optional(),
  primary_language: z.string().default("te-en-IN"),
  persona_template: z.string().default("warm_receptionist"),
});
export type AgentCreate = z.infer<typeof AgentCreate>;

export const AgentUpdate = z.object({
  name: z.string().min(1).max(150).optional(),
  business_identity: z.string().min(1).max(200).optional(),
  description: z.string().nullable().optional(),
  primary_language: z.string().optional(),
});
export type AgentUpdate = z.infer<typeof AgentUpdate>;

export const AgentOut = z.object({
  id: z.string().uuid(),
  name: z.string(),
  business_identity: z.string(),
  description: z.string().nullable(),
  status: z.string(),
  primary_language: z.string(),
  published_version_id: z.string().uuid().nullable(),
  active_phone_number_id: z.string().uuid().nullable(),
  persona_template: z.string().nullable(),
  created_at: z.string(),
});
export type AgentOut = z.infer<typeof AgentOut>;

export const AgentVersionOut = z.object({
  id: z.string().uuid(),
  agent_id: z.string().uuid(),
  version_number: z.number(),
  status: z.string(),
  primary_objective: z.string(),
  ai_disclosure_text: z.string(),
  greeting_text: z.string(),
  closing_text: z.string(),
  personality: z.string(),
  formality: z.string(),
  energy: z.string(),
  response_length: z.string(),
  use_honorifics: z.boolean(),
  supported_languages: z.array(z.string()),
  code_switching_behavior: z.string(),
  restricted_phrases: z.array(z.string()),
  escalation_policy: z.record(z.unknown()),
  quality_score: z.number().nullable(),
  published_at: z.string().nullable(),
  created_at: z.string(),
});
export type AgentVersionOut = z.infer<typeof AgentVersionOut>;

export const AgentDetail = AgentOut.extend({ versions: z.array(AgentVersionOut) });
export type AgentDetail = z.infer<typeof AgentDetail>;

export const AgentVersionUpdate = z.object({
  primary_objective: z.string().optional(),
  ai_disclosure_text: z.string().optional(),
  greeting_text: z.string().optional(),
  closing_text: z.string().optional(),
  personality: z.string().optional(),
  formality: z.string().optional(),
  energy: z.string().optional(),
  response_length: z.string().optional(),
  use_honorifics: z.boolean().optional(),
  supported_languages: z.array(z.string()).optional(),
  code_switching_behavior: z.string().optional(),
  restricted_phrases: z.array(z.string()).optional(),
  escalation_policy: z.record(z.unknown()).optional(),
});
export type AgentVersionUpdate = z.infer<typeof AgentVersionUpdate>;

export const VoicePersonaOut = z.object({
  id: z.string().uuid(),
  provider: z.string(),
  voice_id: z.string(),
  gender_presentation: z.string(),
  language: z.string(),
  speaking_speed: z.number(),
  stability: z.number(),
  expressiveness: z.number(),
  fallback_voice_id: z.string().nullable(),
  sample_audio_url: z.string().nullable(),
});
export type VoicePersonaOut = z.infer<typeof VoicePersonaOut>;

export const VoicePersonaUpdate = z.object({
  provider: z.string().optional(),
  voice_id: z.string().optional(),
  gender_presentation: z.string().optional(),
  language: z.string().optional(),
  speaking_speed: z.number().min(0.5).max(2.0).optional(),
  stability: z.number().min(0).max(1).optional(),
  expressiveness: z.number().min(0).max(1).optional(),
  fallback_voice_id: z.string().nullable().optional(),
});
export type VoicePersonaUpdate = z.infer<typeof VoicePersonaUpdate>;

export const ConversationPolicyOut = z.object({
  id: z.string().uuid(),
  interruption_enabled: z.boolean(),
  min_interruption_ms: z.number(),
  accidental_interruption_phrases: z.array(z.string()),
  silence_timeout_ms: z.number(),
  max_monologue_ms: z.number(),
  max_response_sentences: z.number(),
  confirmation_behavior: z.string(),
  clarification_behavior: z.string(),
  background_noise_tolerance: z.string(),
  human_transfer_enabled: z.boolean(),
  call_later_enabled: z.boolean(),
  wrong_number_behavior: z.string(),
  do_not_call_behavior: z.string(),
});
export type ConversationPolicyOut = z.infer<typeof ConversationPolicyOut>;

export const ConversationPolicyUpdate = z.object({
  interruption_enabled: z.boolean().optional(),
  min_interruption_ms: z.number().min(0).max(5000).optional(),
  accidental_interruption_phrases: z.array(z.string()).optional(),
  silence_timeout_ms: z.number().min(1000).max(60000).optional(),
  max_monologue_ms: z.number().min(1000).max(60000).optional(),
  max_response_sentences: z.number().min(1).max(10).optional(),
  confirmation_behavior: z.string().optional(),
  clarification_behavior: z.string().optional(),
  background_noise_tolerance: z.string().optional(),
  human_transfer_enabled: z.boolean().optional(),
  call_later_enabled: z.boolean().optional(),
  wrong_number_behavior: z.string().optional(),
  do_not_call_behavior: z.string().optional(),
});
export type ConversationPolicyUpdate = z.infer<typeof ConversationPolicyUpdate>;

export const PronunciationEntryOut = z.object({
  id: z.string().uuid(),
  term: z.string(),
  pronunciation: z.string(),
  language: z.string(),
});
export type PronunciationEntryOut = z.infer<typeof PronunciationEntryOut>;

export const PronunciationEntryCreate = z.object({
  term: z.string().min(1).max(200),
  pronunciation: z.string().min(1).max(200),
  language: z.string(),
});
export type PronunciationEntryCreate = z.infer<typeof PronunciationEntryCreate>;

export const AgentVersionDetail = AgentVersionOut.extend({
  voice_persona: VoicePersonaOut.nullable(),
  conversation_policy: ConversationPolicyOut.nullable(),
  pronunciation_entries: z.array(PronunciationEntryOut),
});
export type AgentVersionDetail = z.infer<typeof AgentVersionDetail>;

export const PersonaTemplateOut = z.object({ key: z.string(), label: z.string() });
export type PersonaTemplateOut = z.infer<typeof PersonaTemplateOut>;

export const LANGUAGE_OPTIONS = [
  { value: "te-IN", label: "Telugu" },
  { value: "hi-IN", label: "Hindi" },
  { value: "en-IN", label: "English (India)" },
  { value: "te-en-IN", label: "Telugu-English" },
  { value: "hi-en-IN", label: "Hindi-English" },
] as const;

File: packages/db/jkr_db/models/agents.py
from __future__ import annotations

import uuid
from datetime import datetime

from sqlalchemy import Boolean, Float, ForeignKey, Integer, String, UniqueConstraint
from sqlalchemy.dialects.postgresql import ARRAY, JSONB, UUID
from sqlalchemy.orm import Mapped, mapped_column

from jkr_db.base import Base, TenantMixin
from jkr_db.enums import (
    AgentStatus,
    AgentVersionStatus,
    LanguageCode,
    ProviderName,
    ToolName,
)


class Agent(Base, TenantMixin):
    __tablename__ = "agents"

    name: Mapped[str] = mapped_column(String(150), nullable=False)
    business_identity: Mapped[str] = mapped_column(String(200), nullable=False)
    description: Mapped[str | None] = mapped_column(String(1000), nullable=True)
    status: Mapped[AgentStatus] = mapped_column(String(32), nullable=False, default=AgentStatus.DRAFT)
    primary_language: Mapped[LanguageCode] = mapped_column(
        String(16), nullable=False, default=LanguageCode.TELUGU_ENGLISH
    )
    published_version_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("agent_versions.id", ondelete="SET NULL", use_alter=True, name="fk_agents_published_version"),
        nullable=True,
    )
    active_phone_number_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True), ForeignKey("phone_numbers.id", ondelete="SET NULL"), nullable=True
    )
    persona_template: Mapped[str | None] = mapped_column(String(64), nullable=True)
    # No vocabulary assigned is a normal, valid state — extraction simply
    # skips domain normalization for this agent. See DomainVocabulary below.
    domain_vocabulary_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True), ForeignKey("domain_vocabularies.id", ondelete="SET NULL"), nullable=True
    )


class AgentVersion(Base, TenantMixin):
    __tablename__ = "agent_versions"
    __table_args__ = (UniqueConstraint("agent_id", "version_number", name="uq_agent_version_number"),)

    agent_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("agents.id", ondelete="CASCADE"), nullable=False, index=True
    )
    version_number: Mapped[int] = mapped_column(Integer, nullable=False)
    status: Mapped[AgentVersionStatus] = mapped_column(
        String(32), nullable=False, default=AgentVersionStatus.DRAFT
    )
    primary_objective: Mapped[str] = mapped_column(String(100), nullable=False)
    ai_disclosure_text: Mapped[str] = mapped_column(String(500), nullable=False)
    greeting_text: Mapped[str] = mapped_column(String(1000), nullable=False)
    closing_text: Mapped[str] = mapped_column(String(1000), nullable=False)
    personality: Mapped[str] = mapped_column(String(64), nullable=False, default="warm_receptionist")
    formality: Mapped[str] = mapped_column(String(32), nullable=False, default="balanced")
    energy: Mapped[str] = mapped_column(String(32), nullable=False, default="medium")
    response_length: Mapped[str] = mapped_column(String(32), nullable=False, default="short")
    use_honorifics: Mapped[bool] = mapped_column(Boolean, nullable=False, default=True)
    supported_languages: Mapped[list[str]] = mapped_column(
        ARRAY(String(16)), nullable=False, default=list
    )
    code_switching_behavior: Mapped[str] = mapped_column(
        String(32), nullable=False, default="adaptive"
    )
    restricted_phrases: Mapped[list[str]] = mapped_column(ARRAY(String(200)), nullable=False, default=list)
    escalation_policy: Mapped[dict] = mapped_column(JSONB, nullable=False, default=dict)
    quality_score: Mapped[float | None] = mapped_column(Float, nullable=True)
    published_at: Mapped[datetime | None] = mapped_column(nullable=True)
    created_by: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True), ForeignKey("users.id", ondelete="SET NULL"), nullable=True
    )


class VoicePersona(Base, TenantMixin):
    __tablename__ = "voice_personas"

    agent_version_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("agent_versions.id", ondelete="CASCADE"), nullable=False, unique=True
    )
    provider: Mapped[ProviderName] = mapped_column(String(32), nullable=False, default=ProviderName.MOCK)
    voice_id: Mapped[str] = mapped_column(String(100), nullable=False, default="mock-warm-female-te")
    gender_presentation: Mapped[str] = mapped_column(String(32), nullable=False, default="female")
    language: Mapped[LanguageCode] = mapped_column(String(16), nullable=False, default=LanguageCode.TELUGU_ENGLISH)
    speaking_speed: Mapped[float] = mapped_column(Float, nullable=False, default=1.0)
    stability: Mapped[float] = mapped_column(Float, nullable=False, default=0.6)
    expressiveness: Mapped[float] = mapped_column(Float, nullable=False, default=0.6)
    fallback_voice_id: Mapped[str | None] = mapped_column(String(100), nullable=True)
    sample_audio_url: Mapped[str | None] = mapped_column(String(500), nullable=True)


class PronunciationEntry(Base, TenantMixin):
    __tablename__ = "pronunciation_entries"

    agent_version_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("agent_versions.id", ondelete="CASCADE"), nullable=False, index=True
    )
    term: Mapped[str] = mapped_column(String(200), nullable=False)
    pronunciation: Mapped[str] = mapped_column(String(200), nullable=False)
    language: Mapped[LanguageCode] = mapped_column(String(16), nullable=False)


class ConversationPolicy(Base, TenantMixin):
    __tablename__ = "conversation_policies"

    agent_version_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("agent_versions.id", ondelete="CASCADE"), nullable=False, unique=True
    )
    interruption_enabled: Mapped[bool] = mapped_column(Boolean, nullable=False, default=True)
    min_interruption_ms: Mapped[int] = mapped_column(Integer, nullable=False, default=250)
    accidental_interruption_phrases: Mapped[list[str]] = mapped_column(
        ARRAY(String(64)), nullable=False,
        default=lambda: ["hmm", "okay", "haa", "అవును", "achha", "theek hai", "right"],
    )
    silence_timeout_ms: Mapped[int] = mapped_column(Integer, nullable=False, default=6000)
    max_monologue_ms: Mapped[int] = mapped_column(Integer, nullable=False, default=12000)
    max_response_sentences: Mapped[int] = mapped_column(Integer, nullable=False, default=3)
    confirmation_behavior: Mapped[str] = mapped_column(String(32), nullable=False, default="confirm_critical")
    clarification_behavior: Mapped[str] = mapped_column(String(32), nullable=False, default="ask_dont_guess")
    background_noise_tolerance: Mapped[str] = mapped_column(String(32), nullable=False, default="medium")
    human_transfer_enabled: Mapped[bool] = mapped_column(Boolean, nullable=False, default=True)
    call_later_enabled: Mapped[bool] = mapped_column(Boolean, nullable=False, default=True)
    wrong_number_behavior: Mapped[str] = mapped_column(String(32), nullable=False, default="apologize_and_suppress")
    do_not_call_behavior: Mapped[str] = mapped_column(String(32), nullable=False, default="immediate_suppress")
    # Safety-net ceilings for the shared conversation engine (packages/conversation)
    # — a normal call ends when the objective reaches a terminal state on its
    # own; these only fire if something goes wrong and a call would otherwise
    # run unbounded. Replaces the old hardcoded MAX_AGENT_TURNS=5 module
    # constant in services/api/app/modules/live_call/service.py.
    max_turns: Mapped[int] = mapped_column(Integer, nullable=False, default=30)
    max_call_duration_seconds: Mapped[int] = mapped_column(Integer, nullable=False, default=300)


class DomainVocabulary(Base, TenantMixin):
    """A named set of business-domain terms (dental, admissions, real
    estate, ...) an agent can be assigned to, so extraction can recognize
    when a transcribed value is probably a mis-hearing of a real domain
    term — see jkr_conversation.domain_normalizer. Not agent-exclusive by
    FK direction (Agent -> DomainVocabulary) so one vocabulary can be
    shared across multiple agents in a workspace."""

    __tablename__ = "domain_vocabularies"

    name: Mapped[str] = mapped_column(String(150), nullable=False)
    description: Mapped[str | None] = mapped_column(String(500), nullable=True)


class DomainTerm(Base, TenantMixin):
    __tablename__ = "domain_terms"

    vocabulary_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("domain_vocabularies.id", ondelete="CASCADE"), nullable=False, index=True
    )
    canonical: Mapped[str] = mapped_column(String(200), nullable=False)
    # Alternate spellings/mishearings — no independent lifecycle of their
    # own this pass (no per-alias approval state), so a plain array column
    # rather than a join table, matching restricted_phrases/
    # accidental_interruption_phrases elsewhere in this same file.
    aliases: Mapped[list[str]] = mapped_column(ARRAY(String(200)), nullable=False, default=list)
    category: Mapped[str] = mapped_column(String(64), nullable=False, default="general")
    # "standard" | "critical" — critical terms get confirmed before being
    # trusted under the default confirm_critical policy; see
    # jkr_conversation.engine's confirmation decision table.
    criticality: Mapped[str] = mapped_column(String(16), nullable=False, default="standard")
    languages: Mapped[list[str]] = mapped_column(ARRAY(String(16)), nullable=False, default=list)


class ToolDefinition(Base, TenantMixin):
    __tablename__ = "tool_definitions"
    __table_args__ = (UniqueConstraint("workspace_id", "name", name="uq_tool_definition_workspace_name"),)

    name: Mapped[ToolName] = mapped_column(String(64), nullable=False)
    description: Mapped[str] = mapped_column(String(500), nullable=False)
    input_schema: Mapped[dict] = mapped_column(JSONB, nullable=False, default=dict)
    output_schema: Mapped[dict] = mapped_column(JSONB, nullable=False, default=dict)
    required_permission: Mapped[str] = mapped_column(String(100), nullable=False)
    timeout_seconds: Mapped[int] = mapped_column(Integer, nullable=False, default=10)
    retry_policy: Mapped[dict] = mapped_column(JSONB, nullable=False, default=dict)
    idempotency_behavior: Mapped[str] = mapped_column(String(32), nullable=False, default="key_based")
    confirmation_required: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False)
    audit_enabled: Mapped[bool] = mapped_column(Boolean, nullable=False, default=True)
    failure_fallback: Mapped[str] = mapped_column(String(64), nullable=False, default="offer_human_callback")
    is_enabled: Mapped[bool] = mapped_column(Boolean, nullable=False, default=True)


class AgentTool(Base, TenantMixin):
    __tablename__ = "agent_tools"
    __table_args__ = (
        UniqueConstraint("agent_version_id", "tool_definition_id", name="uq_agent_tool"),
    )

    agent_version_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("agent_versions.id", ondelete="CASCADE"), nullable=False, index=True
    )
    tool_definition_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("tool_definitions.id", ondelete="CASCADE"), nullable=False
    )
    enabled: Mapped[bool] = mapped_column(Boolean, nullable=False, default=True)
    config_override: Mapped[dict] = mapped_column(JSONB, nullable=False, default=dict)
